src/utils/mtev_lfu.c

```diff
@@ -0,0 +1,319 @@
+  }
+}
+
+static bool
+hs_string_compare(const void *a, const void *b)
+{
+  return strcmp((const char * const)a, (const char * const)b) == 0;
+}
+
+mtev_lfu_t *
+mtev_lfu_create(int32_t max_entries, void (*free_fn)(void *))
+{
+  struct mtev_lfu *r = malloc(sizeof(struct mtev_lfu));
+  if (max_entries <= 0) {
+    hs_init(&r->hash, CK_HS_MODE_OBJECT | CK_HS_MODE_DELETE, lfu_entry_hash, hs_string_compare, 1024);
+  } else {
+    hs_init(&r->hash, CK_HS_MODE_OBJECT | CK_HS_MODE_DELETE, lfu_entry_hash, hs_string_compare, max_entries * 2);
+  }
+  STAILQ_INIT(&r->lfu_frequency_list);
+  r->lfu_cache_size = 0;
+  r->expire_count = 0;
+  r->max_entries = max_entries;
+  if (free_fn != NULL) {
+    r->free_fn = free_fn;
+  } else {
+    r->free_fn = free;
+  }
+  pthread_mutex_init(&r->mutex, NULL);
+  return r;
+}
+
+void
+mtev_lfu_destroy(mtev_lfu_t *lfu)
+{
+  mtev_lfu_invalidate(lfu);
+  ck_hs_destroy(&lfu->hash);
+  free(lfu);
+}
+
+void
+mtev_lfu_invalidate(mtev_lfu_t *lfu)
+{
+  pthread_mutex_lock(&lfu->mutex);
+
+  while (!STAILQ_EMPTY(&lfu->lfu_frequency_list)) {
+    struct lfu_cache_entry *e = STAILQ_FIRST(&lfu->lfu_frequency_list);
+    STAILQ_REMOVE(&lfu->lfu_frequency_list, e, lfu_cache_entry, list_entry);
+    while (!STAILQ_EMPTY(&e->lfu_cache)) {
+      struct lfu_entry *le = STAILQ_FIRST(&e->lfu_cache);
+      STAILQ_REMOVE(&e->lfu_cache, le, lfu_entry, freq_list_entry);
+      lfu->free_fn(le->entry);
+      free(le);
+    }
+    free(e);
+  }
+
+  ck_hs_reset(&lfu->hash);
+  lfu->lfu_cache_size = 0;
+  pthread_mutex_unlock(&lfu->mutex);
+}
+
+/* returns the frequency of the list it was in */
+static void
+remove_from_frequency_list_no_lock(mtev_lfu_t *lfu, struct lfu_entry *le)
+{
+  STAILQ_REMOVE(&le->frequency_list_head->lfu_cache, le, lfu_entry, freq_list_entry);
+  if (STAILQ_FIRST(&le->frequency_list_head->lfu_cache) == NULL) {
+    
+    /* we can remove the entire frequency bucket */
+    STAILQ_REMOVE(&lfu->lfu_frequency_list, le->frequency_list_head, lfu_cache_entry, list_entry);
+    free(le->frequency_list_head);
+  }
+  le->frequency_list_head = NULL;
+}
+
+/* this requires a lock to be held elsewhere 
+ * 
+ * There are potentially several items in the same
+ * frequency bucket, we remove the first one
+ */
+static void
+expire_least_lfu_cache_no_lock(mtev_lfu_t *c)
+{
+  /* peek at the first entry in the frequency list */
+  struct lfu_cache_entry *e = STAILQ_FIRST(&c->lfu_frequency_list);
+  struct lfu_entry *le = STAILQ_FIRST(&e->lfu_cache);
+
+  (void)remove_from_frequency_list_no_lock(c, le);
+
+  c->lfu_cache_size--;
+
+  /* remove from hs */
+  unsigned long hash = CK_HS_HASH(&c->hash, lfu_entry_hash, le->key);
+  ck_hs_remove(&c->hash, hash, le->key);
+
+  int ec = ck_pr_load_32(&c->expire_count);
+  if (ec == GC_CADENCE) {
+    c->expire_count = 0;
+    ck_hs_gc(&c->hash, GC_CADENCE, (rand() % ck_hs_count(&c->hash)));
+  }
+  c->expire_count++;
+
+  if (le->entry != NULL) {
```

Comment:
not sure I agree with only calling the free_fn when the entry is non-null... Possible use for this would be to store, say, file-descriptors in a hash, and avoid allocations by storing them as `(void *) (uintptr_t) fd`. `fd == 0` is valid, and we'd want to be able to close it if the free-function invokes `close(2)`.